src/hpke.rs

```diff
@@ -0,0 +1,1046 @@
+//! HPKE draft 12 implementation
+
+use crate::hkdf::KeyType;
+use crate::{aead, agreement, cpu, digest, ec, error, hkdf, hmac, rand};
+use core::convert::TryFrom;
+
+static VERSION: &[u8] = "HPKE-v1".as_bytes();
+const MAX_PUBLIC_KEY_LEN: usize = 1 + (2 * ec::ELEM_MAX_BYTES);
+const MAX_AEAD_KEY_LEN: usize = 32;
+const MAX_DIGEST_LEN: usize = digest::MAX_OUTPUT_LEN;
+
+/// Mode
+#[derive(Debug, Clone, PartialEq)]
+pub enum Mode {
+    /// Base
+    Base = 0x00,
+    /// Psk
+    Psk = 0x01,
+    /// Auth
+    Auth = 0x02,
+    /// AuthPsk
+    AuthPsk = 0x03,
+}
+
+/// KemId
+#[derive(Debug, Clone, PartialEq, Copy)]
+pub enum KemId {
```

Comment:
For future proofing:

```suggestion
#[non_exhaustive]
pub enum KemId {
```

Reply:
`ring` doesn't support an non DH KEMs, and if it did, that would almost certainly imply a version change. Since `ring` is still <1, future proofing here shouldn't practically make sense.